PL2/PL2-parser/src/main/java/nl/tudelft/pl2016gr2/parser/controller/GfaReader.java

```diff
@@ -2,98 +2,237 @@
 
 import nl.tudelft.pl2016gr2.model.Node;
 import nl.tudelft.pl2016gr2.model.OriginalGraph;
+import nl.tudelft.pl2016gr2.thirdparty.testing.utility.TestId;
 
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
 import java.util.ArrayList;
-import java.util.Scanner;
+import java.util.HashMap;
+import java.util.logging.Level;
+import java.util.logging.Logger;
 
 /**
  * This class reads a gfa file.
  *
  * @author Cas
- *
  */
 public class GfaReader {
 
-  private Scanner sc;
-
+  private static final int SHIFT_BY_BASE_10 = 10;
+  private final ArrayList<String> genomes = new ArrayList<>();
+  private final HashMap<Integer, Node> nodes = new HashMap<>();
+  private final String fileName;
+  @TestId(id = "originalGraph")
   private OriginalGraph originalGraph;
-  public final int numNodes;
 
   /**
    * Creates a reader object and reads the gfa data from the filename.
    *
-   * @param filename  the name of the file to be read.
-   * @param graphsize the size of the graph.
+   * @param fileName the name of the file to read.
    */
-  public GfaReader(String filename, int graphsize) {
-    this.numNodes = graphsize;
-    this.sc = new Scanner(GfaReader.class.getClassLoader().getResourceAsStream(filename));
-    originalGraph = new OriginalGraph();
-    prepNodes(originalGraph, numNodes);
-    read();
-
-    sc.close();
-    sc = null;
+  public GfaReader(String fileName) {
+    this.fileName = fileName;
   }
 
   /**
-   * Initialize the nodes of a graph.
+   * Read the GFA file.
    *
-   * @param graph the graph of which to initialize a set amount of node.
-   * @param nodes the amount of nodes to initialize.
+   * @return the read graph.
    */
-  private static void prepNodes(OriginalGraph graph, int nodes) {
-    for (int i = 1; i < nodes + 1; i++) {
-      graph.addNode(new Node(i, 1, new ArrayList<>(), 0));
+  public OriginalGraph read() {
+    if (originalGraph == null) {
+      try {
+        read(fileName);
+      } catch (IOException ex) {
+        Logger.getLogger(GfaReader.class.getName()).log(Level.SEVERE, null, ex);
+      }
+      originalGraph = new OriginalGraph(nodes, 1, genomes);
     }
+    return originalGraph;
   }
 
   /**
-   * Read the actual data.
+   * Read a GFA file.
+   *
+   * @param fileName the name of the GFA file.
    */
-  private void read() {
-    int count = 0;
-    while (sc.hasNext()) {
-      count++;
-      String next = sc.next();
-      switch (next) {
-        case "H":
-          sc.nextLine();
-          break;
-        case "S":
-          sc.nextLine();
-          break;
-        case "L":
-          parseL();
-          break;
-        default:
-          System.out.println("Not S, H or L at: " + count);
+  private void read(String fileName) throws IOException {
```

Comment:
The only thing here is that the parameter is useless. If you change the name of this method to `parse()`, it is both different from `read()` and sticks with the naming of the other methods (`parseEdge()` etc).